src/dotnet/MLSearch.Service/Indexing/ChatIndexerWorker.cs

```diff
@@ -33,28 +73,27 @@ public async Task ExecuteAsync(int _shardIndex, CancellationToken cancellationTo
         // issues in case of re-sharding or new cluster rollouts.
         // It might have some other concurrent worker has updated
         // a cursor. TLDR: prevent stale cursor data.
-        while (!cancellationToken.IsCancellationRequested) {
-            try {
-                var e = await _channel
-                    .Reader
-                    .ReadAsync(cancellationToken)
-                    .ConfigureAwait(false);
-                var result = await dataIndexer.IndexNextAsync(e.Id, cancellationToken).ConfigureAwait(false);
-                if (!result.IsEndReached) {
-                    // Enqueue event to continue indexing.
-                    if (!Trigger.TryWrite(new MLSearch_TriggerChatIndexing(e.Id))) {
-                        Log.LogWarning("Event queue is full: We can't process till this indexing is fully complete.");
-                        while (!result.IsEndReached) {
-                            result = await dataIndexer.IndexNextAsync(e.Id, cancellationToken).ConfigureAwait(false);
-                        }
-                        Log.LogWarning("Event queue is full: exiting an element.");
-                    }
-                }
+
+        // TODO: add more logging and tracing
+        try {
+            var continueIndexing = false;
+            do {
+                var result = await dataIndexer.IndexNextAsync(input.Id, cancellationToken).ConfigureAwait(false);
+                continueIndexing = !(result.IsEndReached || cancellationToken.IsCancellationRequested);
             }
-            catch (Exception e){
-                Log.LogError(e.Message);
-                throw;
+            while (continueIndexing);
+        }
+        catch (OperationCanceledException) {
+            log.LogInformation("Indexing job for chat '{Id}' is cancelled.", input.Id);
+        }
+        catch (Exception e) {
```

Comment:
I don't think eating up all exceptions is a good idea. 
In my opinion this and above should rethrow.

Reply:
There is no other point to observe this exception, as there is no code waiting for this task completion.